internal/agent/device/applications/controller.go

```diff
@@ -0,0 +1,219 @@
+package applications
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/agent/client"
+	"github.com/flightctl/flightctl/internal/agent/device/fileio"
+	"github.com/flightctl/flightctl/pkg/log"
+)
+
+type Controller struct {
+	podman  *client.Podman
+	writer  fileio.Writer
+	manager Manager
+	log     *log.PrefixLogger
+}
+
+func NewController(
+	podman *client.Podman,
+	manager Manager,
+	writer fileio.Writer,
+	log *log.PrefixLogger,
+) *Controller {
+	return &Controller{
+		log:     log,
+		manager: manager,
+		podman:  podman,
+		writer:  writer,
+	}
+}
+
+func (c *Controller) Sync(ctx context.Context, current, desired *v1alpha1.RenderedDeviceSpec) error {
+	c.log.Debug("Syncing device applications")
+	defer c.log.Debug("Finished syncing device applications")
+
+	currentApps, err := parseApps(ctx, c.podman, current)
+	if err != nil {
+		return err
+	}
+
+	desiredApps, err := parseApps(ctx, c.podman, desired)
+	if err != nil {
+		return err
+	}
+
+	// reconcile image based packages
+	if err := c.ensureImages(ctx, currentApps.ImageBased(), desiredApps.ImageBased()); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Controller) ensureImages(ctx context.Context, currentApps, desiredApps []*application[*v1alpha1.ImageApplicationProvider]) error {
+	added, removed, updated, err := diffApps(currentApps, desiredApps)
```

Comment:
need to handle better condition where current=desired on process restart to ensure apps

Reply:
resolved